icicle/src/vec_ops/vec_ops.cu

```diff
@@ -164,4 +189,43 @@ namespace vec_ops {
 
     return CHK_LAST();
   }
+
+  template <typename E>
+  cudaError_t bit_reverse(const E* input, unsigned size, BitReverseConfig& cfg, E* output)
+  {
+    if (size & (size - 1)) THROW_ICICLE_ERR(IcicleError_t::InvalidArgument, "bit_reverse: size must be a power of 2");
+    E* d_output;
+    if (cfg.is_output_on_device) {
+      d_output = output;
+    } else {
+      // allocate output on gpu
+      CHK_IF_RETURN(cudaMallocAsync(&d_output, sizeof(E) * size, cfg.ctx.stream));
+    }
+    unsigned shift = __builtin_clz(size) + 1;
+    unsigned num_blocks = (size + MAX_THREADS_PER_BLOCK - 1) / MAX_THREADS_PER_BLOCK;
+    if (input == output) {
```

Comment:
I think this is interesting only if the input/output are on device.
if your input/output is not on device then anyway you can copy only the input, reverse in place and copy back to the host output. If input and output both on host it doesn't really matter if they are the same or not.